auth.php

```diff
@@ -47,86 +47,44 @@
 
 $currenturl = new moodle_url('/admin/tool/mfa/auth.php');
 
-if (isset($SESSION->tool_mfa_authenticated) && $SESSION->tool_mfa_authenticated) {
+// Check for overall state.
+$state = \tool_mfa\manager::check_status();
+if ($state == \tool_mfa\plugininfo\factor::STATE_PASS) {
+    unset($SESSION->wantsurl);
     redirect(new moodle_url($wantsurl));
+} else if ($state == \tool_mfa\plugininfo\factor::STATE_FAIL) {
+    \tool_mfa\manager::cannot_login();
 }
 
-$userfactors = \tool_mfa\plugininfo\factor::get_active_user_factor_types();
-
-if (count($userfactors) > 0) {
-    $nextfactor = \tool_mfa\plugininfo\factor::get_next_user_factor();
-    $gracemode = false;
-    $factorname = $nextfactor ? $nextfactor->name : null;
-} else {
-    $factorname = null;
-    $gracemode = true;
-}
-
-$form = new login_form($currenturl, array('factor_name' => $factorname, 'grace_mode' => $gracemode));
-if (isset($nextfactor)) {
-    $factor = \tool_mfa\plugininfo\factor::get_factor($factorname);
+// If not pass/fail, check next factor state.
+$factor = \tool_mfa\plugininfo\factor::get_next_user_factor();
+if ($factor->get_state() == \tool_mfa\plugininfo\factor::STATE_FAIL) {
+    // Catch any instant fail state not caught by manager (fallback).
```

Comment:
Is this possible?

Reply:
It is possible, currently only in the case of fallback. The fallback will only be caught here, as it doesnt appear as a factor in check_state, otherwise that would fail everytime. This is for when next_factor returns the fallback factor.

Reply:
Ok so this got me thinking. We set an explicit order to the factors. Should we care that a future factor has failed? If we only check them in the correct order we should never get to it and it shouldn't fail overall

The main use case for a hard fail is an explicit deny of an email or a push notification factor, neither of which you would get until you are up to that factor, so I can't see a down side for the real factors. If you hypothetically had an iprange with a black list, then you'd simply order that factor to be first.